### PR TITLE
Ensure declaring compilation can be located for source symbols created during compilation for EE evaluation.

### LINKED ISSUE #59093: NullReferenceException in AbstractFlowPass evaluating local function with switch statement in EE
Break in C# code and evaluate the following expression in the EE:
```csharp
new Action<int>(x =>
{
    int F(int y)
    {
        switch (y)
        {
            case > 0: return 1;
            case < 0: return -1;
            case 0: return 0;
            default: return 0;
        }
    }
    F(x);
}).Invoke(1)
```

Result: `NullReferenceException` in `AbstractFlowPass<,>.IsTraditionalSwitch()` since `compilation == null`.

>	Microsoft.CodeAnalysis.CSharp.dll!Microsoft.CodeAnalysis.CSharp.AbstractFlowPass<Microsoft.CodeAnalysis.CSharp.ControlFlowPass.LocalState, Microsoft.CodeAnalysis.CSharp.ControlFlowPass.LocalFunctionState>.IsTraditionalSwitch(Microsoft.CodeAnalysis.CSharp.BoundSwitchStatement node) Line 96	C#
...
>	Microsoft.CodeAnalysis.CSharp.dll!Microsoft.CodeAnalysis.CSharp.Binder.BindLocalFunctionStatement.__runAnalysis|872_0(Microsoft.CodeAnalysis.CSharp.BoundBlock block, Microsoft.CodeAnalysis.CSharp.BindingDiagnosticBag blockDiagnostics, ref Microsoft.CodeAnalysis.CSharp.Binder.<>c__DisplayClass872_0 value) Line 605	C#
...

### PR DESCRIPTION
**Pull request by AlekseyTs**
Fixes #59093.
Related to #58198.

### CODE REVIEW

**Comment by AlekseyTs** [comment]
@jcouv, @cston, @dotnet/roslyn-compiler Please review.

File: src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/Symbols/EEMethodSymbol.cs [hunk 1]
```diff
@@ -94,12 +94,17 @@ internal sealed class EEMethodSymbol : MethodSymbol
...
+            sourceMethodTypeParameters = sourceMethod.TypeParameters;
+            allSourceTypeParameters = allSourceTypeParameters.Concat(sourceMethodTypeParameters);
+
             var getTypeMap = new Func<TypeMap>(() => this.TypeMap);
             _typeParameters = sourceMethodTypeParameters.SelectAsArray(
                 (tp, i, arg) => (TypeParameterSymbol)new EETypeParameterSymbol(this, tp, i, getTypeMap),
                 (object)null);
-            _allTypeParameters = container.TypeParameters.Concat(_typeParameters);
-            this.TypeMap = new TypeMap(allSourceTypeParameters, _allTypeParameters);
+            _allTypeParameters = container.TypeParameters.Concat(_typeParameters).Concat(_typeParameters);
```
Comment (chsienki) [inline src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/Symbols/EEMethodSymbol.cs:106]: What is the purpose of the double `Concat`?

Comment (AlekseyTs) [reply src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/Symbols/EEMethodSymbol.cs:106]: > What is the purpose of the double Concat?

We added some new type parameters to ```allSourceTypeParameters``` above, they should be pared with entries in _allTypeParameters. They should be paired with the same ```_typeParameters```.

**Comment by AlekseyTs** [comment]
@jcouv, @cston, https://github.com/orgs/dotnet/teams/roslyn-compiler Please review.

File: src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/CompilationContext.cs [hunk 2]
```diff
@@ -1646,9 +1648,12 @@ private static NamedTypeSymbol GetNonDisplayClassContainer(NamedTypeSymbol type)
                 {
                     if (IsViableSourceMethod(candidateMethod, desiredMethodName, desiredTypeParameters, sourceMethodMustBeInstance))
                     {
+                        MethodSymbol sourceMethod = new EECompilationContextMethod(candidateSubstitutedSourceMethod.DeclaringCompilation!, candidateMethod.OriginalDefinition);
```
Comment (cston) [inline src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/CompilationContext.cs:1651]: >! [](http://example.com/codeflow?start=136&length=1)

Is `!` necessary?

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.PlaceholderLocal.cs [hunk 3]
```diff
@@ -27,6 +27,7 @@ private sealed class PlaceholderLocal : LocalSymbol
             public PlaceholderLocal(Symbol containingSymbol, object identifier, TypeWithAnnotations type)
             {
                 Debug.Assert(identifier != null);
+                Debug.Assert(containingSymbol is null || containingSymbol.DeclaringCompilation is not null);
```
Comment (cston) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.PlaceholderLocal.cs:30]: >Debug.Assert(containingSymbol is null || containingSymbol.DeclaringCompilation is not null); [](http://example.com/codeflow?start=16&length=92)

Could we also assert something similar in `AbstractFlowPass` constructor?
```
Debug.Assert(_symbol is null || _symbol.DeclaringCompilation is not null);
```

File: src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/Symbols/EEMethodSymbol.cs [hunk 4]
```diff
@@ -94,12 +94,17 @@ internal sealed class EEMethodSymbol : MethodSymbol
             var sourceMethodTypeParameters = sourceMethod.TypeParameters;
             var allSourceTypeParameters = container.SourceTypeParameters.Concat(sourceMethodTypeParameters);
```
Comment (cston) [inline src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/Symbols/EEMethodSymbol.cs:95]: Are these assignments needed?

File: src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/CompilationContext.cs (hunk 2 above)
Comment (AlekseyTs) [reply src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/CompilationContext.cs:1651]: > Is ! necessary?

Yes, despite the assert above, a nullable warning is reported here. I will open an issue after the change is merged.

**Review by cston — APPROVED** [review]

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 1)

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.PlaceholderLocal.cs (hunk 3 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.PlaceholderLocal.cs:30]: Unfortunately there are other scenarios (not EE related) when this assert is going to fail. For whatever reason ```AttributeSemanticModel``` uses the attribute type as member symbol and it is used as the symbol for region flow analysis, which makes little sense in my opinion.

**Pull request merged**